File: apps/web/lib/campaigns/schemas.ts
import { z } from 'zod'

/**
 * Zod mirror of `services/api/app/modules/campaigns/domain/models.py::ContentSource`.
 * Keep enum values in sync with the backend Postgres enum.
 */
export const contentSourceSchema = z.enum(['llm', 'edited', 'manual'])
export type ContentSource = z.infer<typeof contentSourceSchema>

/** Zod mirror of `domain/models.py::Priority`. */
export const prioritySchema = z.enum(['high', 'medium', 'low'])
export type Priority = z.infer<typeof prioritySchema>

/** An NPC proposed by the Scribe during extraction — mirrors `schemas.py::ExtractedNPC`. */
export const extractedNpcSchema = z.object({
  name: z.string(),
  description: z.string(),
  current_state: z.string(),
  motivation: z.string(),
  content_source: contentSourceSchema.default('llm'),
})
export type ExtractedNpc = z.infer<typeof extractedNpcSchema>

/** A faction proposed by the Scribe — mirrors `schemas.py::ExtractedFaction`. */
export const extractedFactionSchema = z.object({
  name: z.string(),
  description: z.string(),
  current_stance: z.string(),
  goals: z.string(),
  content_source: contentSourceSchema.default('llm'),
})
export type ExtractedFaction = z.infer<typeof extractedFactionSchema>

/**
 * An arc proposed by the Scribe — mirrors `schemas.py::ExtractedArc`.
 * No `status` field: arc status is assigned on persistence, never proposed.
 */
export const extractedArcSchema = z.object({
  title: z.string(),
  description: z.string(),
  priority: prioritySchema.default('medium'),
  content_source: contentSourceSchema.default('llm'),
})
export type ExtractedArc = z.infer<typeof extractedArcSchema>

/**
 * `POST /campaigns/extract` response body — mirrors `schemas.py::ExtractCampaignOutput`.
 */
export const extractCampaignOutputSchema = z.object({
  title: z.string(),
  description: z.string(),
  world_state: z.string(),
  npcs: z.array(extractedNpcSchema).default([]),
  factions: z.array(extractedFactionSchema).default([]),
  arcs: z.array(extractedArcSchema).default([]),
})
export type ExtractCampaignOutput = z.infer<typeof extractCampaignOutputSchema>

/**
 * `POST /campaigns/extract` request body — the backend trust boundary
 * (mirrors `schemas.py::ExtractRequest`, `Field(min_length=100, max_length=8000)`).
 */
export const extractRequestSchema = z.object({
  raw_text: z
    .string()
    .min(100, 'Premise must be at least 100 characters')
    .max(8000, 'Premise must be at most 8000 characters'),
})
export type ExtractRequest = z.infer<typeof extractRequestSchema>

/** An NPC in a reviewed `POST /campaigns` payload — mirrors `schemas.py::CreateNpcRequest`. */
export const createNpcRequestSchema = z.object({
  name: z.string(),
  description: z.string(),
  current_state: z.string(),
  motivation: z.string(),
  content_source: contentSourceSchema,
})
export type CreateNpcRequest = z.infer<typeof createNpcRequestSchema>

/** A faction in a reviewed payload — mirrors `schemas.py::CreateFactionRequest`. */
export const createFactionRequestSchema = z.object({
  name: z.string(),
  description: z.string(),
  current_stance: z.string(),
  goals: z.string(),
  content_source: contentSourceSchema,
})
export type CreateFactionRequest = z.infer<typeof createFactionRequestSchema>

/**
 * An arc in a reviewed payload — mirrors `schemas.py::CreateArcRequest`.
 * No `status` field accepted from the client — status is always assigned
 * server-side (`"open"`).
 */
export const createArcRequestSchema = z.object({
  title: z.string(),
  description: z.string(),
  priority: prioritySchema.default('medium'),
  content_source: contentSourceSchema,
})
export type CreateArcRequest = z.infer<typeof createArcRequestSchema>

/** `POST /campaigns` request body — mirrors `schemas.py::CreateCampaignRequest`. */
export const createCampaignRequestSchema = z.object({
  title: z.string().min(1),
  description: z.string().min(1),
  world_state: z.string().min(1),
  npcs: z.array(createNpcRequestSchema).default([]),
  factions: z.array(createFactionRequestSchema).default([]),
  arcs: z.array(createArcRequestSchema).default([]),
})
export type CreateCampaignRequest = z.infer<typeof createCampaignRequestSchema>

/** `POST /campaigns` success response body — mirrors `schemas.py::CreateCampaignResponse`. */
export const createCampaignResponseSchema = z.object({
  id: z.string(),
})
export type CreateCampaignResponse = z.infer<
  typeof createCampaignResponseSchema
>

File: services/api/app/modules/campaigns/schemas.py
"""HTTP/LLM contract schemas for the campaigns module.

``ExtractCampaignOutput`` serves both as the ``complete_json`` validation
target and the ``/campaigns/extract`` response model (design Decision 4).
"""

from typing import Any

from pydantic import BaseModel, Field, model_validator

from app.modules.campaigns.domain import ContentSource, Priority


class ScribeExtractedModel(BaseModel):
    """Base for LLM-extracted entities whose provenance is always the Scribe.

    Extraction output is not DM-authored yet. If an LLM hallucinates
    ``content_source: manual`` or ``edited``, normalize it before validation so
    the UI cannot mistake Scribe text for DM-authored text.
    """

    @model_validator(mode="before")
    @classmethod
    def force_scribe_provenance(cls, data: Any) -> Any:
        """Overwrite extracted entity provenance before field validation."""
        if isinstance(data, dict):
            return {**data, "content_source": ContentSource.llm}
        return data


class ExtractedNPC(ScribeExtractedModel):
    """An NPC proposed by the Scribe during extraction."""

    name: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    current_state: str = Field(min_length=1, max_length=1000)
    motivation: str = Field(min_length=1, max_length=1000)
    content_source: ContentSource = ContentSource.llm


class ExtractedFaction(ScribeExtractedModel):
    """A faction proposed by the Scribe during extraction."""

    name: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    current_stance: str = Field(min_length=1, max_length=1000)
    goals: str = Field(min_length=1, max_length=1000)
    content_source: ContentSource = ContentSource.llm


class ExtractedArc(ScribeExtractedModel):
    """An arc proposed by the Scribe during extraction.

    No ``status`` field — arc status is assigned on persistence, never
    proposed by the LLM (see ``campaign-persistence`` spec, CP-003).
    """

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    priority: Priority = Priority.medium
    content_source: ContentSource = ContentSource.llm


class ExtractCampaignOutput(BaseModel):
    """LLM extraction target AND ``POST /campaigns/extract`` response body."""

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    world_state: str = Field(min_length=1, max_length=4000)
    npcs: list[ExtractedNPC] = Field(default_factory=list)
    factions: list[ExtractedFaction] = Field(default_factory=list)
    arcs: list[ExtractedArc] = Field(default_factory=list)


class ExtractRequest(BaseModel):
    """``POST /campaigns/extract`` request body — the backend trust boundary."""

    raw_text: str = Field(min_length=100, max_length=8000)


class CreateNpcRequest(BaseModel):
    """An NPC in a reviewed ``POST /campaigns`` payload."""

    name: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    current_state: str = Field(min_length=1, max_length=1000)
    motivation: str = Field(min_length=1, max_length=1000)
    content_source: ContentSource


class CreateFactionRequest(BaseModel):
    """A faction in a reviewed ``POST /campaigns`` payload."""

    name: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    current_stance: str = Field(min_length=1, max_length=1000)
    goals: str = Field(min_length=1, max_length=1000)
    content_source: ContentSource


class CreateArcRequest(BaseModel):
    """An arc in a reviewed ``POST /campaigns`` payload.

    No ``status`` field accepted from the client — status is always
    assigned by the persistence layer (``"open"``).
    """

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    priority: Priority = Priority.medium
    content_source: ContentSource


class CreateCampaignRequest(BaseModel):
    """``POST /campaigns`` request body — the DM-reviewed payload."""

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=4000)
    world_state: str = Field(min_length=1, max_length=4000)
    npcs: list[CreateNpcRequest] = Field(default_factory=list)
    factions: list[CreateFactionRequest] = Field(default_factory=list)
    arcs: list[CreateArcRequest] = Field(default_factory=list)


class CreateCampaignResponse(BaseModel):
    """``POST /campaigns`` success response body."""

    id: str
